### PR TITLE
Parse lockfile recursively

### PR DESCRIPTION
**Pull request by miiila**
- 💥 Tests written and linted [ℹ︎]() 
- [ ] Documentation written [ℹ︎]()
- [x] Commit history is tidy [ℹ︎]()

### What this does

Introduces a recursive algorithm for creation of dependency tree from `package-lock.json` and `package.json`.

### Notes for the reviewer
I did my best to comment the algorithm, but feel free to suggest any naming or other structural improvements.

### More information

- [SC-5885]()

### CODE REVIEW

File: lib/index.ts [hunk 1]
```diff
@@ -40,46 +42,63 @@ function buildDepTree(targetFileRaw, lockFileRaw, options) {
...
+      // repeat the process for dependencies of looked-up dep
+      if (deps[dep].requires) {
+        Object.keys(deps[dep].requires).forEach(async (subDep) => {
+          depSubTree.dependencies[subDep] = await buildSubTreeRecursive(subDep, [...depKeys, subDep]);
+
+        });
+        return depSubTree;
+      } else {
+        // no more deps, return tree
+        return depSubTree;
```
Comment (lili2311) [inline lib/index.ts]: don't think you need an `else` here since you return in the statement above

File: lib/index.ts [hunk 2]
```diff
@@ -40,46 +42,63 @@ function buildDepTree(targetFileRaw, lockFileRaw, options) {
...
+      }
+    } else {
+      // tree was walked to the root and dependency was not found
+      if (!depKeys.length) {
+        throw new Error(`Dependency ${dep} was not found in package-lock.json.`);
```
Comment (lili2311) [inline lib/index.ts]: Can we make this message more verbose and let the user know that it looks like their package.json and their lockfile are out of sync and give them some advice how to sync them up?

File: lib/index.ts [hunk 3]
```diff
@@ -40,46 +42,63 @@ function buildDepTree(targetFileRaw, lockFileRaw, options) {
...
+
+    // If exists and looked-up dep is there
+    if (deps && deps[dep]) {
+      // update the tree
+      depSubTree.version = deps[dep].version;
+      // repeat the process for dependencies of looked-up dep
+      if (deps[dep].requires) {
+        Object.keys(deps[dep].requires).forEach(async (subDep) => {
+          depSubTree.dependencies[subDep] = await buildSubTreeRecursive(subDep, [...depKeys, subDep]);
+
```
Comment (lili2311) [inline lib/index.ts]: Can you run the linter on this, looks like they might be some linting fixes to add.

File: lib/index.ts (hunk 1 above)
Comment (miiila) [reply lib/index.ts]: You are right, I'll move return to the separate statement.

File: lib/index.ts (hunk 2 above)
Comment (miiila) [reply lib/index.ts]: Will try :-)

File: lib/index.ts (hunk 3 above)
Comment (miiila) [reply lib/index.ts]: `tslint` doesn't complain.

File: package.json [hunk 4]
```diff
@@ -25,11 +25,12 @@
   "devDependencies": {
     "@types/node": "10.5.5",
     "@types/sinon": "5.0.1",
+    "semantic-release": "^8.2.0",
```
Comment (adrukh) [inline package.json]: Please remove :)

File: lib/index.ts (hunk 3 above)
Comment (lili2311) [reply lib/index.ts]: Hah strange!

Comment (miiila) [reply lib/index.ts]: Didn't complain :-D It does now.

File: bin/index.js [hunk 5]
```diff
@@ -0,0 +1,10 @@
+#!/usr/bin/env node
+const inspect = require('../dist/index')
+
+inspect.buildDepTreeFromFiles('./', 'package.json', 'package-lock.json')
+  .then((tree) => {
+    console.log(JSON.stringify(tree));
+  })
+  .catch((e) => {
```
Comment (darscan) [inline bin/index.js]: `.catch(console.error);` might be better here

File: package.json [hunk 6]
```diff
@@ -2,6 +2,9 @@
   "name": "snyk-nodejs-lockfile-parser",
   "description": "Generate a dep tree given a lockfile",
   "main": "dist/lib/index.js",
+  "bin": {
+    "parse": "./bin/index.js"
```
Comment (darscan) [inline package.json]: I think `"parse"` is probably too generic for a bin that will be installed globally

File: lib/index.ts [hunk 7]
```diff
@@ -96,3 +73,25 @@ function getDepPath(depKeys: string[]) {
 
   return depPath;
 }
+
+export function buildDepTreeFromFiles(root, targetFilePath, lockFilePath, options) {
+  if (!root || !lockFilePath || !lockFilePath) {
+    throw new Error('Missing required parameters for parseLockFile()');
+  }
+  // TODO: validate only valid options were passed in
```
Comment (darscan) [inline lib/index.ts]: I might drop this TODO.. just noise

File: package.json (hunk 6 above)
Comment (darscan) [reply package.json]: Not sure what a better name might be.. maybe "parse-npm-lockfile"??

**Pull request merged**